src/error.rs

```diff
@@ -346,9 +330,55 @@ pub(crate) mod private {
 		Custom(String),
 	}
 
+	impl From<std::io::Error> for private::InnerError {
+		fn from(error: std::io::Error) -> Self {
+			private::InnerError::Io(error)
+		}
+	}
+
+	impl From<UnexpectedMessageType> for private::InnerError {
+		fn from(error: UnexpectedMessageType) -> Self {
+			private::InnerError::UnexpectedMessageType(error)
+		}
+	}
+
+	impl std::error::Error for InnerError {}
+
+	impl Display for InnerError {
+		fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+			match self {
+				InnerError::Io(error) => write!(f, "{}", error),
+				InnerError::MessageTooShort { message_len } => write!(
+					f,
+					"the message is too short to be valid: need at least {} for the header, got only {message_len} bytes",
+					crate::HEADER_LEN
+				),
+				InnerError::InvalidMessageType { value } => write!(f, "invalid message type: expected a value in the range [0..4], got {value}"),
+				InnerError::PayloadTooLarge { body_len, max_len } => {
+					write!(f, "payload too large: maximum payload size is {max_len}, got {body_len}")
+				},
+				InnerError::DuplicateRequestId { request_id } => write!(
+					f,
+					"duplicate request ID: request ID {request_id} is already associated with an open request"
+				),
+				InnerError::UnknownRequestId { request_id } => {
+					write!(f, "unknown request ID: request ID {request_id} is not associated with an open request")
+				},
+				InnerError::UnexpectedMessageType(error) => write!(f, "{}", error),
+				InnerError::UnexpectedServiceId { service_id } => write!(f, "unexpected service ID: {service_id}"),
+				InnerError::NoFreeRequestIdFound => write!(f, "no free request ID was found"),
+				InnerError::RequestClosed => write!(f, "the request is already closed"),
+				InnerError::EncodeFailed(error) => write!(f, "{}", error),
+				InnerError::DecodeFailed(error) => write!(f, "{}", error),
+				InnerError::RemoteError(error) => write!(f, "{}", error),
+				InnerError::Custom(error) => write!(f, "{}", error),
+			}
+		}
+	}
+
 	/// Check if a message size is large enough to contain a valid message.
 	#[allow(dead_code)] // not used when all transports are disabled.
-	pub fn check_message_too_short(message_len: usize) -> Result<(), InnerError> {
+	pub fn check_message_too_short(message_len: usize) -> std::result::Result<(), InnerError> {
```

Comment:
```suggestion
	pub fn check_message_too_short(message_len: usize) -> Result<(), InnerError> {
```

This should really just stay `Result`. See also the comment above about import `std::fmt::Result`.